infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientFactory.java

```diff
@@ -63,22 +126,67 @@ public KubernetesClientFactory(
     if (doTrustCerts != null) {
       configBuilder.withTrustCerts(doTrustCerts);
     }
-    this.client = new UnclosableKubernetesClient(configBuilder.build());
+
+    Config config = configBuilder.build();
+    return config;
   }
 
   /**
    * Creates instance of {@link KubernetesClient}.
    *
    * @throws InfrastructureException if any error occurs on client instance creation.
    */
-  public KubernetesClient create() throws InfrastructureException {
-    return client;
+  private KubernetesClient create(Config config) throws InfrastructureException {
+    OkHttpClient clientHttpClient =
+        httpClient.newBuilder().authenticator(Authenticator.NONE).build();
+    OkHttpClient.Builder builder = clientHttpClient.newBuilder();
+    builder.interceptors().clear();
+    clientHttpClient =
+        builder
+            .addInterceptor(
+                new Interceptor() {
+                  @Override
+                  public Response intercept(Chain chain) throws IOException {
+                    Request request = chain.request();
+                    if (isNotNullOrEmpty(config.getUsername())
+                        && isNotNullOrEmpty(config.getPassword())) {
+                      Request authReq =
+                          chain
+                              .request()
+                              .newBuilder()
+                              .addHeader(
+                                  "Authorization",
+                                  Credentials.basic(config.getUsername(), config.getPassword()))
+                              .build();
+                      return chain.proceed(authReq);
+                    } else if (isNotNullOrEmpty(config.getOauthToken())) {
+                      Request authReq =
+                          chain
+                              .request()
+                              .newBuilder()
+                              .addHeader("Authorization", "Bearer " + config.getOauthToken())
+                              .build();
+                      return chain.proceed(authReq);
+                    }
+                    return chain.proceed(request);
+                  }
+                })
+            .build();
+
+    return new UnclosableKubernetesClient(clientHttpClient, config);
   }
 
   @PreDestroy
   private void cleanup() {
     try {
-      client.doClose();
+      if (httpClient.connectionPool() != null) {
+        httpClient.connectionPool().evictAll();
```

Comment:
Please, take a look at my comment to OpenshiftClientFactory

Reply:
Actually looks like you can make this method protected and call it from the OpenshiftClientFactory to avoid duplicated code. WDYT?

Reply:
Good idea.

Reply:
I fixed this in commit https://github.com/eclipse/che/pull/8805/commits/2f9d7cf6e0fc28b099f36180b4c4303569e9ca43 by simply using the same code as `KubernetesClient.close()`, inside a `doCleanup()` protected  method that is used in both `KubernetesClientFactory` and `OpenshiftClientFactory`.